src/games/game.js

```diff
@@ -0,0 +1,123 @@
+   * @param {(a: String, b: String) => String} [scoreSorter]
+   * - Function used to sort scores.
+   * @param {Object} [embedOptions] - Options for the embed message. Defaults to `{ inline: false }`.
+   */
+  constructor(
+    name,
+    messageParser,
+    replyFormatter,
+    scoreFormatter = (score) => score,
+    scoreSorter = (a, b) => a - b,
+    embedOptions = { inline: false, order: undefined },
+  ) {
+    this.name = name
+    this.messageParser = messageParser
+    this.replyFormatter = replyFormatter
+    this.scoreFormatter = scoreFormatter
+    this.scoreSorter = scoreSorter
+    this.embedOptions = embedOptions
+
+    this.topScores = []
+  }
+
+  /**
+   * Attempts to parse a message into a day and score. If successful, an entry is created and
+   * upserted into the database, and a reply is sent to the message.
+   *
+   * @param {Message} message - The message to handle.
+   */
+  async handleMessage(message) {
+    const { day, score } = this.messageParser(message)
+
+    if (day === undefined || score === undefined) {
+      console.debug(`Message is invalid for game ${this.name}`)
+    } else {
+      console.info(`Received valid ${this.name} message`)
+
+      const entry = toEntry(message, this.name, day, score)
+      this.sendReply(message, entry)
+      await upsertEntry(entry)
+    }
+  }
+
+  /**
+   * Sends a reply to a message with a given entry.
+   *
+   * The reply is formatted using the `replyFormatter` function, and the score given to the
+   * `replyFormatter` is formatted using the `scoreFormatter` function.
+   *
+   * @param {Message} message - The message to send a reply to.
+   * @param {Object} entry - The entry to send a reply with.
+   */
+  async sendReply(message, entry) {
+    const score = this.scoreFormatter(entry.score)
+    const reply = this.replyFormatter(entry, score)
+    await message.reply(reply)
+  }
+
+  /**
+   * Updates the top-most scores for the game for the current day.
+   *
+   * If `limit` is provided, only the first `n = limit` scores are saved.
+   *
+   * @param {Number} [limit] - The number of top scores to save.
+   */
+  async updateTopScores(limit) {
+    console.debug(`Updating top scores for game ${this.name}`)
+
+    const today = new Date().toISOString().split('T')[0]
+
+    const entries = await Entry.find({ game: this.name, createdAt: { $gte: today } }).exec()
+    console.debug(`Found ${entries.length} entries for game ${this.name}`)
+
+    const sorted_entries = entries.sort((a, b) => this.scoreSorter(a.score, b.score))
+    this.topScores = sorted_entries.slice(0, limit)
+  }
+
+  /**
+   * Returns a string representation of the top scores for the game.
+   *
+   * Each entry is formatted as "<username> | <score>", with a link to the user's comment.
+   *
+   * @returns {String} A string representation of the top scores for the game.
+   */
+  async getFormattedTopScores() {
+    // TODO: ew?
+    await this.updateTopScores()
```

Comment:
hadde vært nice å få cachet det er og synce en gang i blant (ex. på upsert eller på app startup hvor await er unavoidable), så slipper vi den evig lange promise / await chainen som starter her

Reply:
Er litt vondt å se på men alle spill gjør den entry updaten samtidig når du requester et embed, så går forsåvidt greit tenker eg. Eneste store forbedring eg kan tenke meg er noe caching / conditional pull men det virker komplisert (og litt unødvendig Sd)

Reply:
trenger egentlig ikke conditional pull, bare hent on startup og lagre recieved entries fra disc i cache før upsert, eneste å tenke på det er failure recovery på upsert men det e for edge casey til at eg gidder bry meg